bin/add_print_holdings.rb

```diff
@@ -24,13 +25,43 @@
 # rubocop:enable Layout/LineLength
 
 count = 0
-Zinzout.zin(ARGV.shift).each do |line|
+update = ARGV[0] == "-u"
+if update
+  filename = ARGV[1]
+  logger.info "Updating Holdings."
+else
+  filename = ARGV[0]
+  logger.info "Adding Holdings."
+end
+
+current_date = nil
+organization = nil
+
+Zinzout.zin(filename).each do |line|
   next if /^OCN\tBIB/.match?(line)
   waypoint.incr
   h = Holding.new_from_holding_file_line(line)
-  c = ClusterHolding.new(h).cluster
+
+  current_date = (current_date || h.date_received)
+  organization = (organization || h.organization)
+  c = if update
+    ClusterHolding.new(h).update
```

Comment:
Probably premature optimization, but  it looks like the ClusterHolding service looks up the cluster based on the holding's OCN, but at least in this case we've got the cluster in our hot little hands from line 31. No idea how expensive that lookup is, but might be worth thinking about being able to optionally pass in the cluster at ClusterHolding initialization (unless we want to look it up every time in case something changed underneath?)

Reply:
I think the line numbers have changed since you commented? 

This might be applicable to the delete operation in the update block starting at line 56

Reply:
Right. In general, ClusterHolding takes a holding and the first thing it does for some operations is look up a cluster from the holding. If finding the cluster of a holding is at all expensive, we might want to allow a cluster passing in. 

We should probably have a JIRA ticket/super-ticket for benchmarking / profiling stuff.